In [1]:
import pandas as pd
import numpy as np
import keras
from keras.layers import Dense
from keras.models import Sequential
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_squared_error

Using TensorFlow backend.


In [2]:
concrete_data = pd.read_csv('https://s3-api.us-geo.objectstorage.softlayer.net/cf-courses-data/CognitiveClass/DL0101EN/labs/data/concrete_data.csv')
concrete_data.head()

Cement  Blast Furnace Slag  Fly Ash  Water  Superplasticizer  \
0   540.0                 0.0      0.0  162.0               2.5   
1   540.0                 0.0      0.0  162.0               2.5   
2   332.5               142.5      0.0  228.0               0.0   
3   332.5               142.5      0.0  228.0               0.0   
4   198.6               132.4      0.0  192.0               0.0   

   Coarse Aggregate  Fine Aggregate  Age  Strength  
0            1040.0           676.0   28     79.99  
1            1055.0           676.0   28     61.89  
2             932.0           594.0  270     40.27  
3             932.0           594.0  365     41.05  
4             978.4           825.5  360     44.30

In [3]:
predictors = concrete_data.drop(['Age','Strength'],axis=1) # all columns except Strength
target = concrete_data['Strength']

In [4]:
predictors.head()


Cement  Blast Furnace Slag  Fly Ash  Water  Superplasticizer  \
0   540.0                 0.0      0.0  162.0               2.5   
1   540.0                 0.0      0.0  162.0               2.5   
2   332.5               142.5      0.0  228.0               0.0   
3   332.5               142.5      0.0  228.0               0.0   
4   198.6               132.4      0.0  192.0               0.0   

   Coarse Aggregate  Fine Aggregate  
0            1040.0           676.0  
1            1055.0           676.0  
2             932.0           594.0  
3             932.0           594.0  
4             978.4           825.5

In [5]:
target.head()

0    79.99
1    61.89
2    40.27
3    41.05
4    44.30
Name: Strength, dtype: float64

In [6]:
predictors_norm = (predictors - predictors.mean()) / predictors.std()
predictors_norm.head()

Cement  Blast Furnace Slag   Fly Ash     Water  Superplasticizer  \
0  2.476712           -0.856472 -0.846733 -0.916319         -0.620147   
1  2.476712           -0.856472 -0.846733 -0.916319         -0.620147   
2  0.491187            0.795140 -0.846733  2.174405         -1.038638   
3  0.491187            0.795140 -0.846733  2.174405         -1.038638   
4 -0.790075            0.678079 -0.846733  0.488555         -1.038638   

   Coarse Aggregate  Fine Aggregate  
0          0.862735       -1.217079  
1          1.055651       -1.217079  
2         -0.526262       -2.239829  
3         -0.526262       -2.239829  
4          0.070492        0.647569

In [7]:
n_cols = predictors_norm.shape[1]

In [8]:
#keep 30% data for test 
X_train, X_test, y_train, y_test = train_test_split(predictors, target, test_size=0.30, random_state=42)

# A. Build a baseline model

In [9]:
#Model 
# define regression model
def regression_model():
    # create model
    model = Sequential()
    model.add(Dense(10, activation='relu', input_shape=(n_cols,)))
    model.add(Dense(1))
    
    # compile model
    model.compile(optimizer='adam', loss='mean_squared_error')
    return model

In [10]:
model = regression_model()

Instructions for updating:
Colocations handled automatically by placer.


In [11]:
mselist = []
for i in range (50):
    model.fit(X_train, y_train, validation_split=0.3, epochs=50, verbose=2)
    y_hat = model.predict(X_test)
    merror = mean_squared_error(y_test, y_hat)
    mselist.append(merror)

Instructions for updating:
Use tf.cast instead.
Train on 504 samples, validate on 217 samples
Epoch 1/50
 - 1s - loss: 1611249.8115 - val_loss: 1407176.9810
Epoch 2/50
 - 0s - loss: 1306404.4206 - val_loss: 1135372.8168
Epoch 3/50
 - 0s - loss: 1055796.6071 - val_loss: 918632.3018
Epoch 4/50
 - 0s - loss: 856861.1200 - val_loss: 747295.3828
Epoch 5/50
 - 0s - loss: 699809.5377 - val_loss: 610999.6077
Epoch 6/50
 - 0s - loss: 574668.9921 - val_loss: 501968.5160
Epoch 7/50
 - 0s - loss: 473877.8963 - val_loss: 413879.0000
Epoch 8/50
 - 0s - loss: 391668.9608 - val_loss: 342407.4293
Epoch 9/50
 - 0s - loss: 324916.2986 - val_loss: 283154.4499
Epoch 10/50
 - 0s - loss: 269420.7827 - val_loss: 234577.9371
Epoch 11/50
 - 0s - loss: 223844.5573 - val_loss: 194366.3323
Epoch 12/50
 - 0s - loss: 186059.7793 - val_loss: 161207.3275
Epoch 13/50
 - 0s - loss: 154810.9784 - val_loss: 133950.3335
Epoch 14/50
 - 0s - loss: 129114.8280 - val_loss: 111603.6090
Epoch 15/50
 - 0s - loss: 107966.3255 - va

In [12]:
mse = np.array(mselist)

In [13]:
mse.std()

280.52146583859985

In [15]:
mse.mean()

253.50792821125322

# B Normalize the Data 

In [16]:
# data is already normalized above insted of using predictors, let's use predictors_norm and repeat the process 
X_train, X_test, y_train, y_test = train_test_split(predictors_norm, target, test_size=0.30, random_state=42)

In [17]:
mselist = []
for i in range (50):
    model.fit(X_train, y_train, validation_split=0.3, epochs=50, verbose=2)
    y_hat = model.predict(X_test)
    merror = mean_squared_error(y_test, y_hat)
    mselist.append(merror)

Train on 504 samples, validate on 217 samples
Epoch 1/50
 - 0s - loss: 1599.1329 - val_loss: 1453.8290
Epoch 2/50
 - 0s - loss: 1535.2156 - val_loss: 1388.0391
Epoch 3/50
 - 0s - loss: 1464.0467 - val_loss: 1319.9999
Epoch 4/50
 - 0s - loss: 1389.1782 - val_loss: 1252.7382
Epoch 5/50
 - 0s - loss: 1317.4236 - val_loss: 1184.9381
Epoch 6/50
 - 0s - loss: 1243.5959 - val_loss: 1119.2469
Epoch 7/50
 - 0s - loss: 1173.0639 - val_loss: 1054.8749
Epoch 8/50
 - 0s - loss: 1104.6355 - val_loss: 992.5139
Epoch 9/50
 - 0s - loss: 1038.8163 - val_loss: 933.5522
Epoch 10/50
 - 0s - loss: 976.9690 - val_loss: 877.5601
Epoch 11/50
 - 0s - loss: 919.3030 - val_loss: 824.2389
Epoch 12/50
 - 0s - loss: 864.6260 - val_loss: 774.9857
Epoch 13/50
 - 0s - loss: 813.6594 - val_loss: 728.9369
Epoch 14/50
 - 0s - loss: 766.4504 - val_loss: 685.0976
Epoch 15/50
 - 0s - loss: 722.4784 - val_loss: 644.2580
Epoch 16/50
 - 0s - loss: 681.1928 - val_loss: 606.4657
Epoch 17/50
 - 0s - loss: 643.8714 - val_loss: 570.

In [18]:
mse_norm = np.array(mselist)

In [19]:
mse_norm.std()

10.563024121933214

In [20]:
mse.mean()

253.50792821125322

## part B mean of mse is same as Part A

# C Increase the number of Epochs to 100 

In [21]:
mselist = []
for i in range (50):
    model.fit(X_train, y_train, validation_split=0.3, epochs=100, verbose=2)
    y_hat = model.predict(X_test)
    merror = mean_squared_error(y_test, y_hat)
    mselist.append(merror)

Train on 504 samples, validate on 217 samples
Epoch 1/100
 - 0s - loss: 118.0826 - val_loss: 144.6063
Epoch 2/100
 - 0s - loss: 118.1291 - val_loss: 144.5167
Epoch 3/100
 - 0s - loss: 118.1009 - val_loss: 144.3016
Epoch 4/100
 - 0s - loss: 118.1175 - val_loss: 144.6115
Epoch 5/100
 - 0s - loss: 118.0927 - val_loss: 144.6731
Epoch 6/100
 - 0s - loss: 118.1024 - val_loss: 144.6284
Epoch 7/100
 - 0s - loss: 118.0916 - val_loss: 144.4315
Epoch 8/100
 - 0s - loss: 118.0840 - val_loss: 144.3913
Epoch 9/100
 - 0s - loss: 118.0642 - val_loss: 144.5954
Epoch 10/100
 - 0s - loss: 118.1344 - val_loss: 144.3456
Epoch 11/100
 - 0s - loss: 118.0694 - val_loss: 144.4325
Epoch 12/100
 - 0s - loss: 118.0877 - val_loss: 144.3861
Epoch 13/100
 - 0s - loss: 118.0950 - val_loss: 144.5878
Epoch 14/100
 - 0s - loss: 118.0541 - val_loss: 144.5247
Epoch 15/100
 - 0s - loss: 118.0951 - val_loss: 144.4227
Epoch 16/100
 - 0s - loss: 118.1289 - val_loss: 144.4727
Epoch 17/100
 - 0s - loss: 118.0743 - val_loss: 144

In [22]:

mse_100 = np.array(mselist)

In [23]:
mse_100.std()

0.5032009238890801

In [24]:
mse_100.mean()

159.3763902215916

## Part C mean of mean of squared errors is less than Part B. It significantly decreases (from 253 to 159)

# D. Increase the number of hidden layers

In [25]:
# define regression model
def regression_model():
    # create model
    model = Sequential()
    model.add(Dense(10, activation='relu', input_shape=(n_cols,)))
    model.add(Dense(10, activation='relu'))
    model.add(Dense(10, activation='relu'))
    model.add(Dense(1))
    
    # compile model
    model.compile(optimizer='adam', loss='mean_squared_error')
    return model

In [26]:
model = regression_model()

In [27]:
mselist = []
for i in range (50):
    model.fit(X_train, y_train, validation_split=0.3, epochs=50, verbose=2)
    y_hat = model.predict(X_test)
    merror = mean_squared_error(y_test, y_hat)
    mselist.append(merror)

Train on 504 samples, validate on 217 samples
Epoch 1/50
 - 2s - loss: 1615.4969 - val_loss: 1486.3748
Epoch 2/50
 - 0s - loss: 1594.1321 - val_loss: 1465.5176
Epoch 3/50
 - 0s - loss: 1569.1548 - val_loss: 1439.6583
Epoch 4/50
 - 0s - loss: 1536.6491 - val_loss: 1407.1576
Epoch 5/50
 - 0s - loss: 1495.5454 - val_loss: 1364.7485
Epoch 6/50
 - 0s - loss: 1442.5251 - val_loss: 1310.4891
Epoch 7/50
 - 0s - loss: 1374.9240 - val_loss: 1242.4942
Epoch 8/50
 - 0s - loss: 1292.5475 - val_loss: 1156.5571
Epoch 9/50
 - 0s - loss: 1191.9768 - val_loss: 1054.0765
Epoch 10/50
 - 0s - loss: 1071.2440 - val_loss: 940.0905
Epoch 11/50
 - 0s - loss: 940.4306 - val_loss: 813.6612
Epoch 12/50
 - 0s - loss: 801.2963 - val_loss: 683.3842
Epoch 13/50
 - 0s - loss: 662.4681 - val_loss: 559.7469
Epoch 14/50
 - 0s - loss: 535.5542 - val_loss: 450.6594
Epoch 15/50
 - 0s - loss: 431.5952 - val_loss: 365.5467
Epoch 16/50
 - 0s - loss: 354.0867 - val_loss: 309.7257
Epoch 17/50
 - 0s - loss: 306.7096 - val_loss: 2

In [28]:
mse_3_dense = np.array(mselist)

In [29]:
mse_3_dense.std()

9.811321622852061

In [30]:
mse_100.mean()

159.3763902215916

## Part D mean of mean of squared errors is same as Part C